src/platform/posix/fd.rs

```diff
@@ -12,113 +12,73 @@
 //
 //  0. You just DO WHAT THE FUCK YOU WANT TO.
 
-use std::io::{self, Read, Write};
-use std::os::unix::io::{AsRawFd, IntoRawFd, RawFd};
-
-use crate::error::*;
+use crate::error::{Error, Result};
 use libc::{self, fcntl, F_GETFL, F_SETFL, O_NONBLOCK};
+use std::io;
+use std::os::unix::io::{AsRawFd, IntoRawFd, RawFd};
 
 /// POSIX file descriptor support for `io` traits.
-pub struct Fd(pub RawFd);
+pub(crate) struct Fd {
+    pub(crate) inner: RawFd,
+    close_fd_on_drop: bool,
+}
 
 impl Fd {
-    pub fn new(value: RawFd) -> Result<Self> {
+    pub fn new(value: RawFd, close_fd_on_drop: bool) -> Result<Self> {
         if value < 0 {
             return Err(Error::InvalidDescriptor);
         }
-
-        Ok(Fd(value))
+        Ok(Fd {
+            inner: value,
+            close_fd_on_drop,
+        })
     }
 
     /// Enable non-blocking mode
     pub fn set_nonblock(&self) -> io::Result<()> {
-        match unsafe { fcntl(self.0, F_SETFL, fcntl(self.0, F_GETFL) | O_NONBLOCK) } {
+        match unsafe { fcntl(self.inner, F_SETFL, fcntl(self.inner, F_GETFL) | O_NONBLOCK) } {
             0 => Ok(()),
             _ => Err(io::Error::last_os_error()),
         }
     }
-}
 
-impl Read for Fd {
-    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-        unsafe {
-            let amount = libc::read(self.0, buf.as_mut_ptr() as *mut _, buf.len());
-
-            if amount < 0 {
-                return Err(io::Error::last_os_error());
-            }
-
-            Ok(amount as usize)
+    pub fn read(&self, buf: &mut [u8]) -> io::Result<usize> {
+        let fd = self.as_raw_fd();
+        let amount = unsafe { libc::read(fd, buf.as_mut_ptr() as *mut _, buf.len()) };
+        if amount < 0 {
+            return Err(io::Error::last_os_error());
         }
+        Ok(amount as usize)
     }
 
-    fn read_vectored(&mut self, bufs: &mut [io::IoSliceMut<'_>]) -> io::Result<usize> {
```

Comment:
Why did you eliminate the `read_vectored` implementation with the underlying `readv` syscall? This is a distinct syscall from `read` that uses scatter-gather I/O. This breaks caller expectations and may have a large perf impact in some usages.

Reply:
Well... you can decide whether to add it back or not.